server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -199,9 +199,15 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
 
     @Override
     public RootObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
-        final var mergeResult = MergeResult.build(this, mergeWith, reason, parentBuilderContext);
+        if (mergeWith instanceof RootObjectMapper == false) {
+            MapperErrors.throwObjectMappingConflictError(mergeWith.name());
+        }
```

Comment:
This is an interesting bit: I moved the cast to before the MergeResult#build call, and performed a more accurate check , given that the expectation is that the argument is a RootObjectMapper